File: pixie/evals/trace_helpers.py
"""Convenience functions that extract an ``Evaluable`` from a trace tree.

These are ``from_trace`` callables for use with :func:`~pixie.evals.eval_utils.run_and_evaluate`
and :func:`~pixie.evals.eval_utils.assert_pass`.
"""

from __future__ import annotations

from pixie.instrumentation.spans import LLMSpan
from pixie.storage.evaluable import Evaluable, as_evaluable
from pixie.storage.tree import ObservationNode


def _flatten(nodes: list[ObservationNode]) -> list[ObservationNode]:
    """Recursively flatten a tree of nodes into a flat list."""
    result: list[ObservationNode] = []
    for node in nodes:
        result.append(node)
        result.extend(_flatten(node.children))
    return result


def last_llm_call(trace: list[ObservationNode]) -> Evaluable:
    """Find the ``LLMSpan`` with the latest ``ended_at`` in the trace tree.

    Args:
        trace: The trace tree (list of root ``ObservationNode`` instances).

    Returns:
        An ``LLMSpanEval`` wrapping the most recently ended ``LLMSpan``.

    Raises:
        ValueError: If no ``LLMSpan`` exists in the trace.
    """
    all_nodes = _flatten(trace)
    llm_nodes = [n for n in all_nodes if isinstance(n.span, LLMSpan)]
    if not llm_nodes:
        raise ValueError("No LLMSpan found in the trace")
    llm_nodes.sort(key=lambda n: n.span.ended_at, reverse=True)
    return as_evaluable(llm_nodes[0].span)


def root(trace: list[ObservationNode]) -> Evaluable:
    """Return the first root node's span as ``Evaluable``.

    Args:
        trace: The trace tree (list of root ``ObservationNode`` instances).

    Returns:
        An ``Evaluable`` wrapping the first root node's span.

    Raises:
        ValueError: If the trace is empty.
    """
    if not trace:
        raise ValueError("Trace is empty")
    return as_evaluable(trace[0].span)

File: ts/src/evals/traceHelpers.ts
/**
 * Convenience functions that extract an Evaluable from a trace tree.
 *
 * These are `fromTrace` callables for use with `runAndEvaluate`
 * and `assertPass`.
 */

import type { LLMSpan } from "../instrumentation/spans";
import type { Evaluable } from "../storage/evaluable";
import { asEvaluable } from "../storage/evaluable";
import type { ObservationNode } from "../storage/tree";

// ── Helpers ──────────────────────────────────────────────────────────────────

function flatten(nodes: ObservationNode[]): ObservationNode[] {
  const result: ObservationNode[] = [];
  for (const node of nodes) {
    result.push(node);
    result.push(...flatten(node.children));
  }
  return result;
}

// ── Public API ───────────────────────────────────────────────────────────────

/**
 * Find the LLMSpan with the latest `endedAt` in the trace tree
 * and return it as an Evaluable.
 *
 * @throws if no LLMSpan exists in the trace.
 */
export function lastLlmCall(trace: ObservationNode[]): Evaluable {
  const allNodes = flatten(trace);
  const llmNodes = allNodes.filter(
    (n) => "operation" in n.span
  );
  if (llmNodes.length === 0) {
    throw new Error("No LLMSpan found in the trace");
  }
  llmNodes.sort(
    (a, b) =>
      (b.span as LLMSpan).endedAt.getTime() -
      (a.span as LLMSpan).endedAt.getTime()
  );
  return asEvaluable(llmNodes[0].span);
}

/**
 * Return the first root node's span as Evaluable.
 *
 * @throws if the trace is empty.
 */
export function root(trace: ObservationNode[]): Evaluable {
  if (trace.length === 0) {
    throw new Error("Trace is empty");
  }
  return asEvaluable(trace[0].span);
}
